File: JavaScript/iteration_vs_recursion.js
// The difference between Iteration and Recursion.
// Iteration

function factorial(n) {
  var answer = 1; // 0!

  for (var loop = 1; loop <= n; loop++) {
    answer = answer * loop; // (n-1)! * n
  }
  return answer;
}

var times = 4;
var result = factorial(times);

console.log(`The iteration result of factorial(${times}) is: ${result}`);

//
// Recursion
//

function factorial2(n) {
  if (n == 0) {
    return 1;
  } else {
    return n * factorial2(n - 1);
  }
}

result = factorial2(times);

console.log(`The iteration result of factorial2(${times}) is: ${result}`);

File: Ruby/iteration_vs_recursion.rb
# The difference between Iteration and Recursion.
# 2021-01-18 18:59:41 +0800

input = ARGV[0].to_i
times = input

# =============================
# Iteration （疊代）
# =============================

def factorial(n)
  ans = 1

  (1..n).each do |i|
    ans = ans * i
  end

  ans
end

# The oneliner solution
def factorial(n)
  (1..n).reduce(&:*)
end

result = factorial(times)

puts "The iteration result of factorial(#{times}) is: #{result}"

# =============================
# Recursion （遞迴）
# =============================

def factorial2(n)
  return 1 if n == 0

  n * factorial2(n - 1)
end

result = factorial2(times)

puts "The recursion result of factorial2(#{times}) is: #{result}"
